access/pagerduty/app.go

```diff
@@ -55,337 +69,537 @@ func (a *App) WaitReady(ctx context.Context) (bool, error) {
+			Created:       time.Now(),
+		},
+	}); err != nil {
+		if strings.HasSuffix(err.Error(), "has already reviewed this request") {
+			log.Debug("Already reviewed the request")
+			return nil
+		}
 		return trace.Wrap(err)
 	}
-	log.Infof("Incident %q has been resolved", incidentID)
 
+	log.Info("Successfully submitted a request approval")
 	return nil
 }
 
-func (a *App) tryAutoApproveRequest(ctx context.Context, req access.Request, incidentID string) error {
-	log := logger.Get(ctx)
-
-	if !lib.IsEmail(req.User) {
-		logger.Get(ctx).Warningf("Failed to auto-approve the request: %q does not look like a valid email", req.User)
-		return nil
-	}
+// resolveIncident resolves the notification incident created by plugin if the incident exists.
+func (a *App) resolveIncident(ctx context.Context, reqID string, resolution Resolution) error {
+	var incidentID string
+	ok, err := a.modifyPluginData(ctx, reqID, func(existing *PluginData) (PluginData, bool) {
+		// If plugin data is empty or missing incidentID, we cannot do anything.
+		if existing == nil {
+			return PluginData{}, false
+		}
+		if incidentID = existing.IncidentID; incidentID == "" {
+			return PluginData{}, false
+		}
 
-	user, err := a.bot.GetUserByEmail(ctx, req.User)
-	if err != nil {
-		if trace.IsNotFound(err) {
-			log.WithError(err).Debugf("Failed to auto-approve the request")
-			return nil
+		// If resolution field is not empty then we already resolved the incident before. In this case we just quit.
+		if existing.RequestData.Resolution.Tag != Unresolved {
+			return PluginData{}, false
 		}
-		return err
-	}
 
-	ctx, log = logger.WithFields(ctx, logger.Fields{
-		"pd_user_email": user.Email,
-		"pd_user_name":  user.Name,
+		// Mark incident as resolved.
+		pluginData := *existing
+		pluginData.Resolution = resolution
+		return pluginData, true
 	})
-
-	isOnCall, err := a.bot.IsUserOnCall(ctx, user.ID)
 	if err != nil {
 		return trace.Wrap(err)
 	}
-	if isOnCall {
-		log.Infof("User is now on-call, auto-approving the request")
-		return a.setRequestState(ctx, req.ID, incidentID, user.Email, access.StateApproved)
+	if !ok {
+		logger.Get(ctx).Debug("Failed to resolve the incident: plugin data is missing")
+		return nil
+	}
+
+	ctx, log := logger.WithField(ctx, "pd_incident_id", incidentID)
+	if err := a.pagerduty.ResolveIncident(ctx, incidentID, resolution); err != nil {
+		return trace.Wrap(err)
 	}
+	log.Info("Successfully resolved the incident")
 
-	log.Debug("User is not on call")
 	return nil
 }
 
-func (a *App) getPluginData(ctx context.Context, reqID string) (PluginData, error) {
-	dataMap, err := a.accessClient.GetPluginData(ctx, reqID)
+// modifyPluginData performs a compare-and-swap update of access request's plugin data.
+func (a *App) modifyPluginData(ctx context.Context, reqID string, fn func(data *PluginData) (PluginData, bool)) (bool, error) {
+	var lastErr error
+	for i := 0; i < maxModifyPluginDataTries; i++ {
```

Comment:
I would convert this to a select statement on a ticker inside the for loop to provide some backoff and be nice to the PD API.

Reply:
I don't see how the backoff is useful here. It's a compare-and-swap trial loop, we can just retry immediately.
And in 99.99% cases there will be at most one retry.